File: app/src/main/java/com/poppin/movies/view/SearchActivity.java
package com.poppin.movies.view;

import android.os.Bundle;
import android.support.constraint.ConstraintLayout;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SearchView;
import android.util.Log;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.poppin.movies.R;
import com.poppin.movies.retrofit.MoviesService;
import com.poppin.movies.adapter.MoviesAdapter;
import com.poppin.movies.models.Movies;
import com.poppin.movies.models.RetornoBusca;
import com.poppin.movies.utils.Constants;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class SearchActivity extends AppCompatActivity implements SearchView.OnQueryTextListener, MoviesAdapter.OnItemClickListener {

    private Retrofit retrofit;
    private MoviesService service;

    private TextView txt;
    private TextView txtEmptyView;

    private ConstraintLayout emptyView;
    private ProgressBar progressBar;

    private RetornoBusca retornoBusca;
    private List<Movies> moviesList;
    private RecyclerView rvMovies;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        configRetrofit();
        SearchView sv = findViewById(R.id.svMovies);
        rvMovies = findViewById(R.id.rvMovies);
        emptyView = findViewById(R.id.emptyView);
        progressBar = findViewById(R.id.progressBar);
        txtEmptyView = findViewById(R.id.txtEmptyView);
        sv.setQueryHint("Digite sua busca");
        emptyView(true, R.string.empty_nao_buscou);
        progress(false);
        sv.setOnQueryTextListener(this);
    }

    private void configRetrofit(){
        retrofit = new Retrofit.Builder()
                .baseUrl(Constants.URL_BASE)
                .addConverterFactory(GsonConverterFactory.create())
                .build();

        service = retrofit.create(MoviesService.class);
    }

    private void buscar(String name){
        Call<RetornoBusca> chamada = service.searchMovie(Constants.API_KEY, name, "movie");
        chamada.enqueue(new Callback<RetornoBusca>() {
            @Override
            public void onResponse(Call<RetornoBusca> call, Response<RetornoBusca> response) {
                progress(false);
                if (response.body() != null) {
                    Log.e("Teste", response.body().toString());
                    retornoBusca = response.body();
                    if(retornoBusca.Search != null) {
                        emptyView(false, 0);
                        moviesList = retornoBusca.Search;
                        setupRecycler(moviesList);
                    } else {
                        emptyView(true, R.string.empty_nao_encontrado);
                    }
                }
            }

            @Override
            public void onFailure(Call<RetornoBusca> call, Throwable t) {
                progress(false);
                Log.e("Error", t.getMessage());
            }
        });
    }

    private void emptyView(boolean b, int valor){
        if (b){
            txtEmptyView.setText(valor);
            emptyView.setVisibility(View.VISIBLE);
            rvMovies.setVisibility(View.GONE);
        } else {
            emptyView.setVisibility(View.GONE);
            rvMovies.setVisibility(View.VISIBLE);
        }
    }

    private void progress(boolean b){
        if(b)
            progressBar.setVisibility(View.VISIBLE);
        else
            progressBar.setVisibility(View.INVISIBLE);
    }

    private void setupRecycler(List<Movies> moviesList) {
        rvMovies.setHasFixedSize(false);
        GridLayoutManager layoutManager = new GridLayoutManager(this, 2);
        rvMovies.setLayoutManager(layoutManager);
        MoviesAdapter adapter = new MoviesAdapter(moviesList, this);
        rvMovies.setAdapter(adapter);
        adapter.setClickListener(this);
    }

    @Override
    public boolean onQueryTextSubmit(String text) {
        Log.e("Texto", text);
        progress(true);
        emptyView(false, 0);
        buscar(text);
        return true;
    }

    @Override
    public boolean onQueryTextChange(String s) {
        if (s.equals("")){
            emptyView(true, R.string.empty_nao_buscou);
        }
        return false;
    }



    @Override
    public void onClick(View view, int position, String imdbID) {

    }
}

File: app/src/main/java/com/rodolfo/movies/view/SearchActivity.kt
package com.rodolfo.movies.view

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.TextView
import androidx.appcompat.widget.SearchView
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.RecyclerView
import com.rodolfo.movies.R
import com.rodolfo.movies.view.adapter.MoviesAdapter
import com.rodolfo.movies.base.BaseActivity
import com.rodolfo.movies.models.Movies
import com.rodolfo.movies.utils.Constants
import com.rodolfo.movies.view.viewmodel.SearchViewModel
import org.koin.android.viewmodel.ext.android.viewModel
import org.koin.standalone.KoinComponent

class SearchActivity : BaseActivity(), SearchView.OnQueryTextListener, MoviesAdapter.OnItemClickListener, KoinComponent {

    lateinit var txtEmptyView: TextView

    lateinit var emptyView: ConstraintLayout
    lateinit var loaderFull: ConstraintLayout

    lateinit var rvMovies: RecyclerView

    private val viewModel: SearchViewModel by viewModel()

    override fun onCreate(savedInstanceState: Bundle?) {
        refLayout = R.layout.activity_search
        super.onCreate(savedInstanceState)
        hideKeyboard()
        configViews()
        emptyView(true, R.string.empty_nao_buscou)
        showLoader(loaderFull, false)
        observeChanges()
    }

    private fun observeChanges() {
        viewModel.movies().observe(this, Observer { retornoBusca ->
            showLoader(loaderFull, false)
            retornoBusca.Search?.let { movieList ->
                setupRecycler(movieList.sortedWith(compareByDescending { movie -> movie.Year }))
            } ?: emptyView(true, R.string.empty_nao_encontrado)
        })
    }

    override fun onResume() {
        super.onResume()
        hideKeyboard()
    }

    private fun configViews() {
        val sv = findViewById<SearchView>(R.id.svMovies)

        sv.queryHint = "Digite sua busca"
        rvMovies = findViewById(R.id.rvMovies)
        emptyView = findViewById(R.id.emptyView)
        loaderFull = findViewById(R.id.loaderFull)
        txtEmptyView = findViewById(R.id.txtEmptyView)
        sv.setOnQueryTextListener(this)

    }

    private fun buscar(name: String) {
        viewModel.searchMovie(name)
    }

    private fun emptyView(b: Boolean, valor: Int) {
        if (b) {
            txtEmptyView.setText(valor)
            emptyView.visibility = View.VISIBLE
            rvMovies.visibility = View.GONE
        } else {
            emptyView.visibility = View.GONE
        }
    }

    private fun setupRecycler(moviesList: List<Movies>) {
        rvMovies.setHasFixedSize(false)
        rvMovies.visibility = View.VISIBLE
        val layoutManager = androidx.recyclerview.widget.GridLayoutManager(this, 2)
        rvMovies.layoutManager = layoutManager
        val adapter = MoviesAdapter(moviesList, this)
        rvMovies.adapter = adapter
        adapter.setClickListener(this)
    }

    override fun onQueryTextSubmit(text: String): Boolean {
        showLoader(loaderFull, true)
        rvMovies.visibility = View.GONE
        emptyView(false, 0)
        hideKeyboard()
        buscar(text)
        return true
    }

    override fun onQueryTextChange(s: String): Boolean {
        if (s == "") {
            emptyView(true, R.string.empty_nao_buscou)
        }
        return true
    }


    override fun onClick(view: View, position: Int, imdbID: String?) {
        val i = Intent(this, DetailsActivity::class.java)
        i.putExtra(Constants.IMDB_ID, imdbID)
        startActivity(i)
    }
}
